anndata/tests/test_raw.py

```diff
@@ -31,37 +31,49 @@
 )
 
 
+@pytest.fixture
+def adata_raw():
+    adata = ad.AnnData(
+        np.array(data), obs=obs_dict, var=var_dict, uns=uns_dict, dtype='int32'
+    )
+    adata.raw = adata
+    return adata[:, [0, 1]]
+
+
 # -------------------------------------------------------------------------------
 # The test functions
 # -------------------------------------------------------------------------------
 
 
-def test_raw(backing_h5ad):
-    X = np.array(X_list)
-    adata = ad.AnnData(
-        X, obs=obs_dict, var=var_dict, uns=uns_dict, dtype='int32'
-    )
-
-    # init raw
-    adata.raw = adata
+def test_raw_init(adata_raw):
+    assert adata_raw.var_names.tolist() == ['var1', 'var2']
+    assert adata_raw.raw.var_names.tolist() == ['var1', 'var2', 'var3']
+    assert adata_raw.raw[:, 0].X.tolist() == [[1], [4], [7]]
 
-    assert adata.raw[:, 0].X.tolist() == [[1], [4], [7]]
 
-    adata = adata[:, [0, 1]]
+def test_raw_del(adata_raw):
+    del adata_raw.raw
+    assert adata_raw.raw is None
 
-    assert adata.var_names.tolist() == ['var1', 'var2']
-    assert adata.raw.var_names.tolist() == ['var1', 'var2', 'var3']
 
-    # read write
+def test_raw_rw(adata_raw, backing_h5ad):
     with pytest.warns(
         ImplicitModificationWarning, match="Initializing view as actual"
     ):  # TODO: don’t modify adata just to write it
-        adata.write(backing_h5ad)
-    adata = ad.read(backing_h5ad)
+        adata_raw.write(backing_h5ad)
+    adata_raw = ad.read(backing_h5ad)
+
+    assert adata_raw.var_names.tolist() == ['var1', 'var2']
+    assert adata_raw.raw.var_names.tolist() == ['var1', 'var2', 'var3']
+    assert adata_raw.raw[:, 0].X.tolist() == [[1], [4], [7]]
+
 
-    assert adata.raw[:, 0].X.tolist() == [[1], [4], [7]]
-    assert adata.raw.var_names.tolist() == ['var1', 'var2', 'var3']
-    assert adata.var_names.tolist() == ['var1', 'var2']
+def test_raw_backed(adata_raw, backing_h5ad):
+    with pytest.warns(
+        ImplicitModificationWarning, match="Initializing view as actual"
+    ):  # TODO: don’t modify adata just to write it
+        adata_raw.filename = backing_h5ad
 
-    del adata.raw
-    assert adata.raw is None
+    assert adata_raw.var_names.tolist() == ['var1', 'var2']
+    assert adata_raw.raw.var_names.tolist() == ['var1', 'var2', 'var3']
+    assert adata_raw.raw[:, 0].X[:].tolist() == [[1], [4], [7]]
```

Comment:
Well, this is broken because of

https://github.com/theislab/anndata/blob/4baf74bed7bf120176d9fa1eb3e59c646ca16a0d/anndata/_core/raw.py#L102-L105

And has been since the beginning. Wat do.

Reply:
How is this broken exactly?

Reply:
`adata_raw.raw[:, 0].X.shape != (3, 1)`

Reply:
Ah, I think that just need to be fixed here:
https://github.com/theislab/anndata/blob/4baf74bed7bf120176d9fa1eb3e59c646ca16a0d/anndata/_core/raw.py#L37-L63

I think this is the need for "view of raw" that pops up in all the comments.

Reply:
If it's always been broken you could probably just open an issue and not touch it for now.

Reply:
definitely out of scope for this PR